spring-xd-yarn/spring-xd-yarn-client/src/main/java/org/springframework/xd/yarn/XdYarnAdminInfoCommand.java

```diff
@@ -0,0 +1,126 @@
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.xd.yarn;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import org.apache.zookeeper.KeeperException;
+import org.apache.zookeeper.WatchedEvent;
+import org.apache.zookeeper.Watcher;
+import org.apache.zookeeper.ZooKeeper;
+import org.apache.zookeeper.data.Stat;
+import org.springframework.beans.factory.config.YamlPropertiesFactoryBean;
+import org.springframework.boot.cli.command.AbstractCommand;
+import org.springframework.boot.cli.command.status.ExitStatus;
+import org.springframework.boot.cli.util.Log;
+import org.springframework.core.io.FileSystemResource;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Properties;
+
+/**
+ * A custom command for listing the http ports used by admin servers.
+ *
+ * @author Thomas Risberg
+ */
+public class XdYarnAdminInfoCommand extends AbstractCommand {
+
+	private static final String XD_YARN_ADMIN_INFO_COMMAND_NAME = "admininfo";
+	private static final String XD_YARN_ADMIN_INFO_COMMAND_DESC = "List http host and port info for the admin servers";
+
+	public static final String ZK_NAMESPACE = "zk.namespace";
+	public static final String ZK_CLIENT_CONNECT = "zk.client.connect";
+
+	public static final String XD_ADMINS = "admins";
+
+	private String zkNamespace = "xd";
+	private String zkClientConnect = "localhost:2181";
+
+	private ZooKeeper zk;
+
+	/**
+	 * Call super constructor for new {@link org.springframework.boot.cli.command.AbstractCommand} instance.
+	 */
+	public XdYarnAdminInfoCommand() {
+		super(XD_YARN_ADMIN_INFO_COMMAND_NAME, XD_YARN_ADMIN_INFO_COMMAND_DESC);
+	}
+
+	@Override
+	public ExitStatus run(String... args) throws InterruptedException {
+		try {
+			init();
+			Log.info("Admins: " + getAdmins());
+		}
+		catch (IOException e) {
+			Log.error("IOException accessing admin info: " + e.getMessage());
+		}
+		catch (KeeperException e) {
+			Log.error("KeeperException accessing admin info: " + e.getMessage());
+		}
+		finally {
+			close();
+		}
+		return ExitStatus.OK;
+	}
+
+	private void init() throws IOException {
+		YamlPropertiesFactoryBean ypfb = new YamlPropertiesFactoryBean();
+		ypfb.setResources(new FileSystemResource(System.getProperty("spring.config.location")));
+		Properties conf = ypfb.getObject();
+		if (conf.containsKey(ZK_NAMESPACE)) {
+			zkNamespace = (String) conf.get(ZK_NAMESPACE);
+		}
+		if (conf.containsKey(ZK_CLIENT_CONNECT)) {
+			zkClientConnect = (String) conf.get(ZK_CLIENT_CONNECT);
+		}
+		zk = new ZooKeeper(zkClientConnect, 5000,
+				new Watcher() {
```

Comment:
Ahh OK, I see the reason why. According to the JavaDoc:

> Session establishment is asynchronous. This constructor will initiate
> connection to the server and return immediately - potentially (usually)
> before the session is fully established. The watcher argument specifies
> the watcher that will be notified of any changes in state. This
> notification can come at any point before or after the constructor call
> has returned.

So that `Watcher` can be used to determine when a session is established. I don't think you'll need it in this case; your read request should simply queue up if it is issued before the session is established.